### ISSUE TITLE
Incorrect timing of gdnative_singleton

### ISSUE THREAD

**Issue by BastiaanOlij** [godotengine/godot#52960]
### Godot version

3.x 1bbc7c9c3a323832e352fe4d74c0da923f0b27ae

### System information

Any

### Issue description

The current startup order of Godot results in an issue that we aren't able to access core parts of the system when we're building a plugin the implement singletons needing to be instantiated at startup.

I'm working on changes in the way ARVR plugins register their interface that raised this issue. 
In the old model we call a dedicated GDNative entry point that instantiates `ARVRInterfaceGDNative`, assigns an entry point struct to it, and registers it with the `ARVRServer`.
In the new model we subclass `ARVRInterfaceGDNative` in our plugin and attempt to register an instance of that class.

While `ARVRServer` has been created at this point in time and it's singleton is accessible, it is not accessibly from GDNative because it hasn't been registered with the engine.

We can follow the code in `main.cpp` for the method `Main::setup2`. We find that:
- on line 1314 we create our `ARVRServer` singleton
- on line 1454 we call `register_module_types` which initialised GDnative and calls our singleton entry point
- on line 1472 we call `register_server_singletons` which registers the `ARVRServer` singleton

So we can see that our registration is too late. We need to bring that forward but `register_server_singletons` also registers singletons that aren't available until later.

I can think of three options:
1) remove registering `ARVRServer` from `register_server_singletons` and call it right after ARVRServer is created
2) move registering `ARVRServer` from `register_server_singletons` into `register_core_singletons`
3) move setting up our gdnative singletons until after `register_server_singletons` was called

### Steps to reproduce

Create a simple GDNative plugin, implement the `..._gdnative_singleton` entry point and attempt to access any of the server singletons.

### Minimal reproduction project

https://github.com/GodotVR/godot_openxr/pull/70 attempts to implement this for the OpenXR plugin, see `godot_openxr.cpp`

**Comment by BastiaanOlij** [comment]
Note that while my problem is with ARVRServer, the same applies for the other server singletons, some which indeed haven't been setup at all by the time GDNative is initialised while i can imagine some GDNative plugins wanting to access those. Option #3 might be the most sensible one but is also the most difficult one.

**Comment by akien-mga** [comment]
I think option 3. makes sense, you'd expect e.g. the PhysicsServer or VisualServer singletons to be properly registered before you attempt to register a GDNative implementation of their APIs (rhetorically speaking, I don't think that's possible for those two in 3.x).

But yeah I'm not sure how to implement this cleanly, for the editor we have `EditorNode::add_init_callback(editor_init_callback);` (used in `register_gdnative_types()`) which can be used to defer registration to after the editor has been fully initialized. Something similar for the engine proper could be useful here, where you could have `Engine::add_init_callback(engine_init_callback)` or similar (or in `Main::`?) that would register the callbacks that `main.cpp` would call when appropriate (e.g. just before `_start_success = true`, where we seem to actually have a similar system for VisualServer callbacks with an ad-hoc solution).

**Comment by BastiaanOlij** [comment]
Ok, after discussing this with @reduz the conclusion for this part of the problem is to register XRServer earlier in the process. In both 3.x and 4.0 we'll move registering XRServer into `register_server_types`.

We're leaving other server singletons where they are. In 3.x nobody currently uses GDnative in this way so its unlikely to be a problem. In 4.0 we've enhanced the registration system so we have more control over when what happens.

**Comment by BastiaanOlij** [comment]
Closing this issue, this has had many discussions over the years, always coming to the same conclusion that there is a misunderstanding between registering features in GDNative/GDExtension, the correct placement of initialisation code, and the needs of core parts of the system.
The direct issue I had back when we worked on OpenXR has been solved differently.

There is still a discussion ongoing whether GDNative/GDExtension accessing singletons the way it is now instead of calling the actual `get_singleton` function, which is possible now in GDExtension, should be kept. But that is outside of the scope of this issue.

Bottom line, decision was made many times that we do not want to change the order or implementation of our initialisation steps
